### PR TITLE
[Speculators][Speculative Decoding] Support gpt-oss eagle3 on blackwell

### PR DESCRIPTION
**Pull request by jiahanc**
## Purpose
Support eagle3 for gpt-oss on Blackwell flashinfer trtllm-gen attention.
This PR uses some content in https://github.com/vllm-project/vllm/pull/25196, hold off till https://github.com/vllm-project/vllm/pull/25196 merged

## Test Plan
lm_eval
```
lm_eval --model local-completions --tasks gsm8k --model_args model=openai/gpt-oss-120b,base_url=http://0.0.0.0:30000/v1/completions,max_retries=3,tokenized_requests=False,timeout=1200,max_gen_toks=2048,max_length=8192 --batch_size 2048 --trust_remote_code --limit 0.8
```
## Test Result
W/ eagle3
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.8286|±  |0.0116|
|     |       |strict-match    |     5|exact_match|↑  |0.6212|±  |0.0149|
```

w/o eagle3
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.8277|±  |0.0116|
|     |       |strict-match    |     5|exact_match|↑  |0.6155|±  |0.0150|
```
## (Optional) Documentation Update

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by jiahanc** [comment]
Performance benchmark on `vllm bench serve` with `spec_bench` dataset concurrency 1 and 100 requests. Seeing 1.5x speed up.
w/o eagle3
```
============ Serving Benchmark Result ============
Successful requests:                     100       
Maximum request concurrency:             1         
Benchmark duration (s):                  415.52    
Total input tokens:                      34733     
Total generated tokens:                  102400    
Request throughput (req/s):              0.24      
Output token throughput (tok/s):         246.44    
Peak output token throughput (tok/s):    251.00    
Peak concurrent requests:                2.00      
Total Token throughput (tok/s):          330.03    
---------------Time to First Token----------------
Mean TTFT (ms):                          23.89     
Median TTFT (ms):                        19.70     
P99 TTFT (ms):                           47.92     
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          4.04      
Median TPOT (ms):                        4.03      
P99 TPOT (ms):                           4.23      
---------------Inter-token Latency----------------
Mean ITL (ms):                           4.04      
Median ITL (ms):                         4.04      
P99 ITL (ms):                            4.40      
----------------End-to-end Latency----------------
Mean E2EL (ms):                          4155.07   
Median E2EL (ms):                        4141.16   
P99 E2EL (ms):                           4348.93   
==================================================
```

w/ eagle3
```
============ Serving Benchmark Result ============
Successful requests:                     100       
Maximum request concurrency:             1         
Benchmark duration (s):                  268.29    
Total input tokens:                      34733     
Total generated tokens:                  102400    
Request throughput (req/s):              0.37      
Output token throughput (tok/s):         381.67    
Total Token throughput (tok/s):          511.13    
---------------Time to First Token----------------
Mean TTFT (ms):                          28.04     
Median TTFT (ms):                        21.66     
P99 TTFT (ms):                           52.68     
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          2.60      
Median TPOT (ms):                        2.59      
P99 TPOT (ms):                           3.34      
---------------Inter-token Latency----------------
Mean ITL (ms):                           6.35      
Median ITL (ms):                         6.35      
P99 ITL (ms):                            6.64      
----------------End-to-end Latency----------------
Mean E2EL (ms):                          2682.78   
Median E2EL (ms):                        2698.46   
P99 E2EL (ms):                           3439.56   
==================================================
```

File: vllm/model_executor/models/llama_eagle3.py [hunk 1]
```diff
@@ -204,6 +204,11 @@ def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
         nn.Module.__init__(self)
         self.config = vllm_config. \
             speculative_config.draft_model_config.hf_config
+        # Ensure draft_vocab_size is set
```
Comment (benchislett) [inline vllm/model_executor/models/llama_eagle3.py:207]: Could you check if this logic is also present in llama_eagle.py?

Comment (jiahanc) [reply vllm/model_executor/models/llama_eagle3.py:207]: Add same logic to `llama_eagle.py`

File: vllm/v1/spec_decode/eagle.py [hunk 2]
```diff
@@ -888,6 +880,30 @@ def validate_same_kv_cache_group(self,
             ])
         ) == 1, "All eagle layers should belong to the same kv cache group"
 
+    def _get_attention_metadata_builder(self, ubatch_id):
```
Comment (benchislett) [inline vllm/v1/spec_decode/eagle.py]: Why is this logic needed?

Comment (jiahanc) [reply vllm/v1/spec_decode/eagle.py]: some models like gpt-oss has multiple attention backends ( sliding window attn and full attn), we want to make sure the draft_model is using the correct attention_metadata builder. Otherwise there will be accuracy issue

**Comment by jiahanc** [comment]
hold off and wait https://github.com/vllm-project/vllm/pull/25109 merge first

File: vllm/v1/spec_decode/eagle.py [hunk 3]
```diff
@@ -191,11 +182,11 @@ def propose(
 
         assert self.runner is not None
 
-        # FIXME: need to consider multiple kv_cache_groups
+        # Select the correct attention metadata builder for EAGLE layers.
         ubatch_id = dbo_current_ubatch_id()
-        attn_metadata_builder = \
-            self.runner.attn_groups[0][0].metadata_builders[ubatch_id]
-        attn_metadata = attn_metadata_builder.build_for_drafting(
+        builder = self._get_attention_metadata_builder(ubatch_id)
```
Comment (heheda12345) [inline vllm/v1/spec_decode/eagle.py]: I think the result of this function can be cached.

Comment (jiahanc) [reply vllm/v1/spec_decode/eagle.py]: Done, cached attention_metadata_builders because different ubatch_id may have diff builder

**Comment by jiahanc** [comment]
this PR is split into 2 PR:
1. Main change to https://github.com/vllm-project/vllm/pull/25406
2. left change are part of https://github.com/vllm-project/vllm/pull/25196
close this one

**Comment by nidhishgajjar** [comment]
**Orb Code Review** (powered by GLM 5.1 on [Orb Cloud](https://orbcloud.dev))

This is a substantial PR adding EAGLE3 speculative decoding support for gpt-oss on Blackwell with FlashInfer/TRT-LLM attention. The changes are well-structured overall. Here are my observations:

**Architecture: ClassVar → Instance Variable for `reorder_batch_threshold`**

The change from `ClassVar[int]` to instance-level `int` across all attention backends is the right call. It enables dynamic threshold computation based on speculative config at construction time via `_init_reorder_batch_threshold()`. The systematic application across all backends (FlashInfer, xformers, FlashAttn, MLA, GDN, Linear, Mamba, ShortConv) is consistent.

**GptOssModel EAGLE3 support:**

- `get_eagle3_aux_hidden_state_layers()` returns `(2, num_layers // 2, num_layers - 3)` — follows the same pattern as other models. Good.
- The `aux_hidden_state_layers` is initialized as an empty `tuple()` and must be set via `set_aux_hidden_state_layers`. This matches the `SupportsEagle3` interface contract, but consider logging a warning if forward runs without it being set (the empty tuple means silently no aux states, which could be confusing to debug).
- Return type change to `Union[torch.Tensor, tuple[torch.Tensor, list[torch.Tensor]]]` is correct but impacts all callers. The conditional `if len(aux_hidden_states) > 0` means it returns a different type at runtime — callers must handle this.

**EagleProposer lm_head sharing logic change:**

The old code excluded eagle3 from lm_head sharing (`method != "eagle3"`). The new code shares based on shape comparison. This is a **behavioral change for eagle3 models** — they will now share lm_head when shapes match. The test results suggest this works, but please confirm this was intentional and tested for all supported eagle3 target models (llama, qwen), not just gpt_oss.

**`split_decodes_and_prefills` with `require_uniform`:**

The logic `is_prefill = query_lens != query_lens[0]` uses the first request as the reference length. This is correct given the batch is pre-ordered (decodes first), but the early-return path at line ~710:
```python
if max_query_len <= decode_threshold and \
    (not require_uniform or decode_threshold <= 1):
```
When `require_uniform=True` and `decode_threshold > 1`, this still returns all-decodes if `max_query_len <= decode_threshold`. This should be fine since uniform decode with qlen ≤ threshold is a valid uniform batch.

**Minor items:**

1. **`config/__init__.py`** — Only whitespace change (trailing newline). Consider dropping this from the PR to reduce noise.
2. **`config/model.py`** — The `quant_method` fix (`if != "" then None`) appears to be a separate bug fix. Worth noting in the PR description or splitting out.
3. **Missing newline at end of file** in `config/__init__.py`, `flashinfer.py`, and `utils.py`. These create noisy diffs in future PRs.
4. **PR depends on #25196** — Noted in the description. Reviewers should verify that dependency is addressed.

**Testing:** The PR includes lm_eval results showing quality preservation with eagle3, which is great. However, for an architectural change of this scope (8+ attention backends modified, new spec-as-decode kernel path), unit tests for `split_decodes_and_prefills` with `require_uniform=True` and for `reshape_query_for_spec_decode` would add confidence.

**Summary:** Well-structured PR adding EAGLE3/gpt-oss support with a clean architectural improvement to attention backend configuration. The lm_head sharing logic change needs confirmation that it's safe for all eagle3 targets. Minor cleanup items (whitespace, separate bug fixes) should be addressed.

**Assessment: comment** — Functionally sound with test results. A few items to clarify/address before merge.
